### PR TITLE
SHIPA-828 Add Platform CRUD

### PR DESCRIPTION
**Pull request by murphybytes**
# Description

This PR adds a Platform CRD and adds CLI commands to add, list and delete platforms.  See [SHIPA-828](https://shipaio.atlassian.net/browse/SHIPA-828).

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Chore (documentation addition or typo, file relocation)

## Testing

- [ ] New tests were added with this PR that prove my fix is effective or that my feature works (describe below this bullet)
- [ ] This change requires no testing (i.e. documentation update)

## Documentation

- [x] All added public packages, funcs, and types have been documented with doc comments
- [x] I have commented my code, particularly in hard-to-understand areas

## Final Checklist:

- [ ] I followed standard [GitHub flow](https://guides.github.com/introduction/flow/) guidelines
- [ ] I have performed a self-review of my own code
- [ ] My changes generate no new warnings

## Additional Information (omit if empty)

Please include anything else relevant to this PR that may be useful to know.

### CODE REVIEW

File: cmd/ketch/platform_add.go [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+	"fmt"
+	"io"
+
+	"github.com/spf13/cobra"
+)
+
+const (
+	platformAddHelp = `
+Adds a new platform.  The platform image can be inferred if you're using an
+official platform. See https://github.com/shipa-corp/platforms for a list.
```
Comment (aleksej-paschenko) [inline cmd/ketch/platform_add.go:15]: shipa-corp/platforms is a private repo as for now

File: cmd/ketch/platform_list_test.go [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+	"github.com/stretchr/testify/require"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/shipa-corp/ketch/internal/api/v1beta1"
+)
+
+func TestPlatformList(t *testing.T) {
+	expected := `+------+-----------------------+-------------+
```
Comment (aleksej-paschenko) [inline cmd/ketch/platform_list_test.go:18]: should we start using it? `app list` and `pool list` use a style similar to kubectl. This type of table data is hard to parse and I'm not sure how to consume this kind of output in scripts

File: internal/api/v1beta1/platform_types.go [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+}
+
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:scope=Cluster
+// +kubebuilder:printcolumn:name="Name",type=string,JSONPath=`.metadata.name`
+// +kubebuilder:printcolumn:name="Image",type=string,JSONPath=`.spec.image`
+// +kubebuilder:printcolumn:name="Description",type=string,JSONPath=`.spec.description`
+
+// Platform an image that is used to build an app. See https://learn.shipa.io/docs/platforms-1
+type Platform struct {
```
Comment (aleksej-paschenko) [inline internal/api/v1beta1/platform_types.go:27]: CRD for platforms, awesome

File: internal/chart/platform_chart.go [hunk 4]
```diff
@@ -0,0 +1,7 @@
+package chart
+
+type platform struct {
```
Comment (aleksej-paschenko) [inline internal/chart/platform_chart.go:3]: not used

File: cmd/ketch/platform_add_test.go [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+				return
+			}
+			require.Nil(t, err)
+			require.Equal(t, tc.called, cli.called)
+		})
+	}
+}
+
+type paCreateMock struct {
+	f      func(o runtime.Object) error
```
Comment (aleksej-paschenko) [inline cmd/ketch/platform_add_test.go:106]: because it's a mock for `resourceCreator`, do you think we can move it to a separate file and rename to `resourceCreatorMock` then we can use it with other commands?

File: cmd/ketch/platform.go [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+	"github.com/spf13/cobra"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/shipa-corp/ketch/internal/api/v1beta1"
+)
+
+func newPlatformCmd(cfg config, logOut io.Writer) *cobra.Command {
```
Comment (aleksej-paschenko) [inline cmd/ketch/platform.go:16]: can we use `out` instead of `logOut`, just for consistency with existing code?

File: cmd/ketch/platform_add.go [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+`
+	imageFlag            = "image"
+	imageShortFlag       = "i"
+	dockerfileFlag       = "dockerfile"
+	dockerfileShortFlag  = "d"
+	descriptionFlag      = "description"
+	descriptionShortFlag = "D"
+)
+
+func newPlatformAddCmd(creator resourceCreator, logWriter io.Writer) *cobra.Command {
```
Comment (aleksej-paschenko) [inline cmd/ketch/platform_add.go:32]: can we have `out` instead of `logWriter`?

File: cmd/ketch/platform_add.go [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+	}
+	return nil
+}
+
+// TODO: Implement see https://shipaio.atlassian.net/browse/SHIPA-831?atlOrigin=eyJpIjoiYjJjMzc4NzJhOWZmNGNmMGIyMDk3YzhlODk1NTUzZjgiLCJwIjoiaiJ9
+func platformFromDockerfile(ctx context.Context, dockerFileReference string) (string, error) {
+	return "", errors.New("build platform from dockerfile not implemented")
+}
+
+func platformAdd(ctx context.Context, creator resourceCreator, options platformAddOptions, logWriter io.Writer) error {
```
Comment (aleksej-paschenko) [inline cmd/ketch/platform_add.go:73]: can we have `out` instead of `logWriter`?

File: cmd/ketch/platform_delete.go [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+
+import (
+	"context"
+	"fmt"
+	"io"
+
+	"github.com/spf13/cobra"
+)
+
+func newPlatformDeleteCmd(cli resourceGetDeleter, logWriter io.Writer) *cobra.Command {
```
Comment (aleksej-paschenko) [inline cmd/ketch/platform_delete.go:11]: can we have `out` instead of `logWriter`?

File: cmd/ketch/platform_list.go [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+	"context"
+	"fmt"
+	"io"
+	"strings"
+	"text/tabwriter"
+
+	"github.com/spf13/cobra"
+)
+
+func newPlatformListCmd(lister resourceLister, logWriter io.Writer) *cobra.Command {
```
Comment (aleksej-paschenko) [inline cmd/ketch/platform_list.go:13]: can we have `out` instead of `logWriter`?

File: cmd/ketch/platform_list.go [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+}
+
+func writePlatformList(ctx context.Context, lister resourceLister, out io.Writer) error {
+	platforms, err := platformList(ctx, lister)
+	if err != nil {
+		return fmt.Errorf("could not list platforms: %w", err)
+	}
+	tw := tabwriter.NewWriter(out, 0, 4, 4, ' ', 0)
+	fmt.Fprintln(tw, "NAME\tIMAGE\tDESCRIPTION")
+	for _, platform := range platforms.Items {
```
Comment (aleksej-paschenko) [inline cmd/ketch/platform_list.go:31]: should we sort platforms before printing?

**Review by aleksej-paschenko — COMMENTED** [review]
looks awesome
I know it is a bit annoying but can we use `out` instead of `logWriter`?

**Review by aleksej-paschenko — APPROVED** [review]

**Comment by vivek** [comment]
platform is no longer supported, closing it.
